{"nbformat":4,"nbformat_minor":0,"metadata":{"accelerator":"GPU","colab":{"name":"second_trainer_counting.ipynb","provenance":[],"collapsed_sections":[],"toc_visible":true,"machine_shape":"hm","mount_file_id":"1m8bh6G3ZCq_xCw17eY9wX8822yxrl-b_","authorship_tag":"ABX9TyN97G+VDURra/iWwtzyhm+u"},"kernelspec":{"name":"python3","display_name":"Python 3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.7.3"}},"cells":[{"cell_type":"code","metadata":{"colab_type":"code","executionInfo":{"status":"ok","timestamp":1587931376644,"user_tz":-120,"elapsed":3645,"user":{"displayName":"horst hammer","photoUrl":"","userId":"16387219538505432860"}},"id":"spDiWJ97I0vZ","outputId":"b302d70b-71bc-4b07-94c4-52f4da685fd0","colab":{"base_uri":"https://localhost:8080/","height":302}},"source":["# %cd ..\n","!nvidia-smi"],"execution_count":2,"outputs":[{"output_type":"stream","text":["Sun Apr 26 20:02:55 2020       \n","+-----------------------------------------------------------------------------+\n","| NVIDIA-SMI 440.64.00    Driver Version: 418.67       CUDA Version: 10.1     |\n","|-------------------------------+----------------------+----------------------+\n","| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |\n","| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |\n","|===============================+======================+======================|\n","|   0  Tesla P100-PCIE...  Off  | 00000000:00:04.0 Off |                    0 |\n","| N/A   47C    P0    28W / 250W |      0MiB / 16280MiB |      0%      Default |\n","+-------------------------------+----------------------+----------------------+\n","                                                                               \n","+-----------------------------------------------------------------------------+\n","| Processes:                                                       GPU Memory |\n","|  GPU       PID   Type   Process name                             Usage      |\n","|=============================================================================|\n","|  No running processes found                                                 |\n","+-----------------------------------------------------------------------------+\n"],"name":"stdout"}]},{"cell_type":"code","metadata":{"colab_type":"code","id":"j5KSeQKwuZYB","outputId":"bdff8710-1321-437c-bbd6-e80274076e4b","executionInfo":{"status":"ok","timestamp":1587931463249,"user_tz":-120,"elapsed":89918,"user":{"displayName":"horst hammer","photoUrl":"","userId":"16387219538505432860"}},"colab":{"base_uri":"https://localhost:8080/","height":1000}},"source":["!pip install gdown\n","!pip install tensorflow-gpu --user\n","!pip install --upgrade keras"],"execution_count":3,"outputs":[{"output_type":"stream","text":["Requirement already satisfied: gdown in /usr/local/lib/python3.6/dist-packages (3.6.4)\n","Requirement already satisfied: six in /usr/local/lib/python3.6/dist-packages (from gdown) (1.12.0)\n","Requirement already satisfied: requests in /usr/local/lib/python3.6/dist-packages (from gdown) (2.21.0)\n","Requirement already satisfied: tqdm in /usr/local/lib/python3.6/dist-packages (from gdown) (4.38.0)\n","Requirement already satisfied: urllib3<1.25,>=1.21.1 in /usr/local/lib/python3.6/dist-packages (from requests->gdown) (1.24.3)\n","Requirement already satisfied: chardet<3.1.0,>=3.0.2 in /usr/local/lib/python3.6/dist-packages (from requests->gdown) (3.0.4)\n","Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.6/dist-packages (from requests->gdown) (2020.4.5.1)\n","Requirement already satisfied: idna<2.9,>=2.5 in /usr/local/lib/python3.6/dist-packages (from requests->gdown) (2.8)\n","Collecting tensorflow-gpu\n","\u001b[?25l  Downloading https://files.pythonhosted.org/packages/0a/93/c7bca39b23aae45cd2e85ad3871c81eccc63b9c5276e926511e2e5b0879d/tensorflow_gpu-2.1.0-cp36-cp36m-manylinux2010_x86_64.whl (421.8MB)\n","\u001b[K     |████████████████████████████████| 421.8MB 40kB/s \n","\u001b[?25hRequirement already satisfied: numpy<2.0,>=1.16.0 in /usr/local/lib/python3.6/dist-packages (from tensorflow-gpu) (1.18.3)\n","Requirement already satisfied: opt-einsum>=2.3.2 in /usr/local/lib/python3.6/dist-packages (from tensorflow-gpu) (3.2.1)\n","Collecting tensorflow-estimator<2.2.0,>=2.1.0rc0\n","\u001b[?25l  Downloading https://files.pythonhosted.org/packages/18/90/b77c328a1304437ab1310b463e533fa7689f4bfc41549593056d812fab8e/tensorflow_estimator-2.1.0-py2.py3-none-any.whl (448kB)\n","\u001b[K     |████████████████████████████████| 450kB 35.4MB/s \n","\u001b[?25hRequirement already satisfied: scipy==1.4.1; python_version >= \"3\" in /usr/local/lib/python3.6/dist-packages (from tensorflow-gpu) (1.4.1)\n","Requirement already satisfied: keras-applications>=1.0.8 in /usr/local/lib/python3.6/dist-packages (from tensorflow-gpu) (1.0.8)\n","Requirement already satisfied: six>=1.12.0 in /usr/local/lib/python3.6/dist-packages (from tensorflow-gpu) (1.12.0)\n","Requirement already satisfied: protobuf>=3.8.0 in /usr/local/lib/python3.6/dist-packages (from tensorflow-gpu) (3.10.0)\n","Collecting tensorboard<2.2.0,>=2.1.0\n","\u001b[?25l  Downloading https://files.pythonhosted.org/packages/d9/41/bbf49b61370e4f4d245d4c6051dfb6db80cec672605c91b1652ac8cc3d38/tensorboard-2.1.1-py3-none-any.whl (3.8MB)\n","\u001b[K     |████████████████████████████████| 3.9MB 42.9MB/s \n","\u001b[?25hRequirement already satisfied: absl-py>=0.7.0 in /usr/local/lib/python3.6/dist-packages (from tensorflow-gpu) (0.9.0)\n","Requirement already satisfied: astor>=0.6.0 in /usr/local/lib/python3.6/dist-packages (from tensorflow-gpu) (0.8.1)\n","Requirement already satisfied: grpcio>=1.8.6 in /usr/local/lib/python3.6/dist-packages (from tensorflow-gpu) (1.28.1)\n","Collecting gast==0.2.2\n","  Downloading https://files.pythonhosted.org/packages/4e/35/11749bf99b2d4e3cceb4d55ca22590b0d7c2c62b9de38ac4a4a7f4687421/gast-0.2.2.tar.gz\n","Requirement already satisfied: wheel>=0.26; python_version >= \"3\" in /usr/local/lib/python3.6/dist-packages (from tensorflow-gpu) (0.34.2)\n","Requirement already satisfied: termcolor>=1.1.0 in /usr/local/lib/python3.6/dist-packages (from tensorflow-gpu) (1.1.0)\n","Requirement already satisfied: wrapt>=1.11.1 in /usr/local/lib/python3.6/dist-packages (from tensorflow-gpu) (1.12.1)\n","Requirement already satisfied: keras-preprocessing>=1.1.0 in /usr/local/lib/python3.6/dist-packages (from tensorflow-gpu) (1.1.0)\n","Requirement already satisfied: google-pasta>=0.1.6 in /usr/local/lib/python3.6/dist-packages (from tensorflow-gpu) (0.2.0)\n","Requirement already satisfied: h5py in /usr/local/lib/python3.6/dist-packages (from keras-applications>=1.0.8->tensorflow-gpu) (2.10.0)\n","Requirement already satisfied: setuptools in /usr/local/lib/python3.6/dist-packages (from protobuf>=3.8.0->tensorflow-gpu) (46.1.3)\n","Requirement already satisfied: google-auth-oauthlib<0.5,>=0.4.1 in /usr/local/lib/python3.6/dist-packages (from tensorboard<2.2.0,>=2.1.0->tensorflow-gpu) (0.4.1)\n","Requirement already satisfied: markdown>=2.6.8 in /usr/local/lib/python3.6/dist-packages (from tensorboard<2.2.0,>=2.1.0->tensorflow-gpu) (3.2.1)\n","Requirement already satisfied: requests<3,>=2.21.0 in /usr/local/lib/python3.6/dist-packages (from tensorboard<2.2.0,>=2.1.0->tensorflow-gpu) (2.21.0)\n","Requirement already satisfied: werkzeug>=0.11.15 in /usr/local/lib/python3.6/dist-packages (from tensorboard<2.2.0,>=2.1.0->tensorflow-gpu) (1.0.1)\n","Requirement already satisfied: google-auth<2,>=1.6.3 in /usr/local/lib/python3.6/dist-packages (from tensorboard<2.2.0,>=2.1.0->tensorflow-gpu) (1.7.2)\n","Requirement already satisfied: requests-oauthlib>=0.7.0 in /usr/local/lib/python3.6/dist-packages (from google-auth-oauthlib<0.5,>=0.4.1->tensorboard<2.2.0,>=2.1.0->tensorflow-gpu) (1.3.0)\n","Requirement already satisfied: chardet<3.1.0,>=3.0.2 in /usr/local/lib/python3.6/dist-packages (from requests<3,>=2.21.0->tensorboard<2.2.0,>=2.1.0->tensorflow-gpu) (3.0.4)\n","Requirement already satisfied: idna<2.9,>=2.5 in /usr/local/lib/python3.6/dist-packages (from requests<3,>=2.21.0->tensorboard<2.2.0,>=2.1.0->tensorflow-gpu) (2.8)\n","Requirement already satisfied: urllib3<1.25,>=1.21.1 in /usr/local/lib/python3.6/dist-packages (from requests<3,>=2.21.0->tensorboard<2.2.0,>=2.1.0->tensorflow-gpu) (1.24.3)\n","Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.6/dist-packages (from requests<3,>=2.21.0->tensorboard<2.2.0,>=2.1.0->tensorflow-gpu) (2020.4.5.1)\n","Requirement already satisfied: rsa<4.1,>=3.1.4 in /usr/local/lib/python3.6/dist-packages (from google-auth<2,>=1.6.3->tensorboard<2.2.0,>=2.1.0->tensorflow-gpu) (4.0)\n","Requirement already satisfied: pyasn1-modules>=0.2.1 in /usr/local/lib/python3.6/dist-packages (from google-auth<2,>=1.6.3->tensorboard<2.2.0,>=2.1.0->tensorflow-gpu) (0.2.8)\n","Requirement already satisfied: cachetools<3.2,>=2.0.0 in /usr/local/lib/python3.6/dist-packages (from google-auth<2,>=1.6.3->tensorboard<2.2.0,>=2.1.0->tensorflow-gpu) (3.1.1)\n","Requirement already satisfied: oauthlib>=3.0.0 in /usr/local/lib/python3.6/dist-packages (from requests-oauthlib>=0.7.0->google-auth-oauthlib<0.5,>=0.4.1->tensorboard<2.2.0,>=2.1.0->tensorflow-gpu) (3.1.0)\n","Requirement already satisfied: pyasn1>=0.1.3 in /usr/local/lib/python3.6/dist-packages (from rsa<4.1,>=3.1.4->google-auth<2,>=1.6.3->tensorboard<2.2.0,>=2.1.0->tensorflow-gpu) (0.4.8)\n","Building wheels for collected packages: gast\n","  Building wheel for gast (setup.py) ... \u001b[?25l\u001b[?25hdone\n","  Created wheel for gast: filename=gast-0.2.2-cp36-none-any.whl size=7540 sha256=630a71292a17c09063adbe98ade564557945e30e9d424ccc74f7e3aad5990398\n","  Stored in directory: /root/.cache/pip/wheels/5c/2e/7e/a1d4d4fcebe6c381f378ce7743a3ced3699feb89bcfbdadadd\n","Successfully built gast\n","\u001b[31mERROR: tensorflow 2.2.0rc3 has requirement gast==0.3.3, but you'll have gast 0.2.2 which is incompatible.\u001b[0m\n","\u001b[31mERROR: tensorflow 2.2.0rc3 has requirement tensorboard<2.3.0,>=2.2.0, but you'll have tensorboard 2.1.1 which is incompatible.\u001b[0m\n","\u001b[31mERROR: tensorflow 2.2.0rc3 has requirement tensorflow-estimator<2.3.0,>=2.2.0rc0, but you'll have tensorflow-estimator 2.1.0 which is incompatible.\u001b[0m\n","\u001b[31mERROR: tensorflow-probability 0.10.0rc0 has requirement gast>=0.3.2, but you'll have gast 0.2.2 which is incompatible.\u001b[0m\n","Installing collected packages: tensorflow-estimator, tensorboard, gast, tensorflow-gpu\n","\u001b[33m  WARNING: The script tensorboard is installed in '/root/.local/bin' which is not on PATH.\n","  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.\u001b[0m\n","\u001b[33m  WARNING: The scripts estimator_ckpt_converter, saved_model_cli, tensorboard, tf_upgrade_v2, tflite_convert, toco and toco_from_protos are installed in '/root/.local/bin' which is not on PATH.\n","  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.\u001b[0m\n","Successfully installed gast-0.2.2 tensorboard-2.1.1 tensorflow-estimator-2.1.0 tensorflow-gpu-2.1.0\n","Requirement already up-to-date: keras in /usr/local/lib/python3.6/dist-packages (2.3.1)\n","Requirement already satisfied, skipping upgrade: keras-preprocessing>=1.0.5 in /usr/local/lib/python3.6/dist-packages (from keras) (1.1.0)\n","Requirement already satisfied, skipping upgrade: keras-applications>=1.0.6 in /usr/local/lib/python3.6/dist-packages (from keras) (1.0.8)\n","Requirement already satisfied, skipping upgrade: scipy>=0.14 in /usr/local/lib/python3.6/dist-packages (from keras) (1.4.1)\n","Requirement already satisfied, skipping upgrade: pyyaml in /usr/local/lib/python3.6/dist-packages (from keras) (3.13)\n","Requirement already satisfied, skipping upgrade: six>=1.9.0 in /usr/local/lib/python3.6/dist-packages (from keras) (1.12.0)\n","Requirement already satisfied, skipping upgrade: h5py in /usr/local/lib/python3.6/dist-packages (from keras) (2.10.0)\n","Requirement already satisfied, skipping upgrade: numpy>=1.9.1 in /usr/local/lib/python3.6/dist-packages (from keras) (1.18.3)\n"],"name":"stdout"}]},{"cell_type":"code","metadata":{"id":"33KexW3M7xhZ","colab_type":"code","outputId":"8f90c841-6031-4baf-f306-cd79e1deb9d6","executionInfo":{"status":"ok","timestamp":1587931468373,"user_tz":-120,"elapsed":93276,"user":{"displayName":"horst hammer","photoUrl":"","userId":"16387219538505432860"}},"colab":{"base_uri":"https://localhost:8080/","height":235}},"source":["#quickfix error in callbacks of keras, pullrequest is already open in github\n","#later versions of tensorflow and keras should contain this fix\n","%cd ../../../../../../\n","from google.colab import files\n","import shutil\n","%cd root/.local/lib/python3.6/site-packages/tensorflow_core/python/keras/\n","# files.download('callbacks.py')\n","!rm callbacks.py\n","shutil.copyfile('/content/drive/My Drive/person_detection/utils/callbacks.py', \n","                './callbacks.py')\n","!ls\n","\n"],"execution_count":4,"outputs":[{"output_type":"stream","text":["/\n","/root/.local/lib/python3.6/site-packages/tensorflow_core/python/keras\n","activations.py\t   estimator\t\t\t   optimizers.py\n","api\t\t   initializers.py\t\t   optimizer_v2\n","applications\t   __init__.py\t\t\t   premade\n","backend_config.py  keras_parameterized.py\t   preprocessing\n","backend.py\t   layers\t\t\t   __pycache__\n","callbacks.py\t   losses.py\t\t\t   regularizers.py\n","callbacks_v1.py    metrics.py\t\t\t   saving\n","constraints.py\t   mixed_precision\t\t   testing_utils.py\n","datasets\t   models.py\t\t\t   utils\n","distribute\t   model_subclassing_test_util.py  wrappers\n","engine\t\t   ops.py\n"],"name":"stdout"}]},{"cell_type":"markdown","metadata":{"colab_type":"text","id":"J8LEtNbNn7Ri"},"source":["# Training\n","\n"]},{"cell_type":"code","metadata":{"colab_type":"code","id":"A-DO3mGwqq0i","colab":{}},"source":["from datetime import datetime\n","\n","logdir = '/content/drive/My\\ Drive/person_counting/tensorboard/colab_{}_trainer_nr_2/'.format(datetime.now().strftime(\"%Y-%m-%d-%H-%M-%S\"))\n","top_path = '/content/drive/My\\ Drive/person_detection/bus_videos/pcds_dataset_detected/'"],"execution_count":0,"outputs":[]},{"cell_type":"code","metadata":{"id":"do7gLVqi_jNB","colab_type":"code","outputId":"3767d042-05f7-4300-dbe1-620d53c6e977","executionInfo":{"status":"ok","timestamp":1587931468376,"user_tz":-120,"elapsed":89051,"user":{"displayName":"horst hammer","photoUrl":"","userId":"16387219538505432860"}},"colab":{"base_uri":"https://localhost:8080/","height":34}},"source":["%cd /content/drive/My\\ Drive/"],"execution_count":6,"outputs":[{"output_type":"stream","text":["/content/drive/My Drive\n"],"name":"stdout"}]},{"cell_type":"code","metadata":{"id":"E8ep37lYAEkX","colab_type":"code","colab":{}},"source":["!python -m person_counting.models.cnn_regression --topdir-log {logdir} \\\n","                                                 --n-runs 100 \\\n","                                                 --top-path {top_path}\\\n","                                                 --batch-size 16\\\n","                                                 --filter-rows-factor 1\\\n","                                                 --filter-cols-lower 30\\\n","                                                 --filter-cols-upper 20\\"],"execution_count":0,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"NQFwhx_FWzGV","colab_type":"text"},"source":["# Testing for best params\n","\n","### Set static params in test file"]},{"cell_type":"code","metadata":{"id":"I2bdEgA9VZUp","colab_type":"code","outputId":"7f626716-ce33-4ec3-8182-a969b6a1e9a2","executionInfo":{"status":"ok","timestamp":1587892718138,"user_tz":-120,"elapsed":431185,"user":{"displayName":"horst hammer","photoUrl":"","userId":"16387219538505432860"}},"colab":{"base_uri":"https://localhost:8080/","height":742}},"source":["#Plotting config\n","%matplotlib inline\n","from pandas.plotting import register_matplotlib_converters\n","import seaborn as sns\n","register_matplotlib_converters()\n","%matplotlib inline\n","sns.set_context('notebook')\n","%config InlineBackend.figure_format = 'retina'\n","\n","%run -m person_counting.tests.test_cnn train_best_gpu"],"execution_count":0,"outputs":[{"output_type":"stream","text":["Dataset contains: \n","0 training csvs \n","0 testing csvs\n","5 4 (1, ['y_stride'])\n","Model: \"sequential_2\"\n","_________________________________________________________________\n","Layer (type)                 Output Shape              Param #   \n","=================================================================\n","conv2d_3 (Conv2D)            (None, 262, 27, 5)        85        \n","_________________________________________________________________\n","max_pooling2d_2 (MaxPooling2 multiple                  0         \n","_________________________________________________________________\n","conv2d_4 (Conv2D)            (None, 128, 10, 5)        405       \n","_________________________________________________________________\n","flatten_2 (Flatten)          (None, 1600)              0         \n","_________________________________________________________________\n","dense_2 (Dense)              (None, 1)                 1601      \n","=================================================================\n","Total params: 2,091\n","Trainable params: 2,091\n","Non-trainable params: 0\n","_________________________________________________________________\n","Actual model is using following hyper-parameters:\n","pooling_type :  max\n","learning_rate :  0.0055113\n","filter_rows_factor :  2\n","filter_cols_factor :  1\n","layer_number :  2\n","pool_size_x :  2\n","pool_size_y :  2\n","kernel_size :  4\n","regularization :  0.05\n","kernel_number :  5\n","y_stride :  1\n","filter_cols_upper :  30\n","filter_cols_lower :  20\n","optimizer :  Adam\n","batch_normalization :  False\n","batch_size :  8\n","Epoch 1/15\n"],"name":"stdout"},{"output_type":"stream","text":["/usr/local/lib/python3.6/dist-packages/keras/engine/training_generator.py:49: UserWarning: Using a generator with `use_multiprocessing=True` and multiple workers may duplicate your data. Please consider using the `keras.utils.Sequence class.\n","  UserWarning('Using a generator with `use_multiprocessing=True`'\n"],"name":"stderr"}]},{"cell_type":"code","metadata":{"id":"kOfBHA8FTVdB","colab_type":"code","colab":{}},"source":["# use this code to prevent colab from disconnecting during longer training times in javascript console window\n","function ClickConnect(){\n","    console.log(\"Clicked on connect button\"); \n","    document.querySelector(\"colab-connect-button\").click()\n","}\n","setInterval(ClickConnect,60000)"],"execution_count":0,"outputs":[]}]}